docs: add apisix-gitops-adc.md (#1703)
iresty/iresty.github.io: blog/en/blog/2023/10/08/apisix-gitops-adc.md

<!--truncate-->

With the widespread adoption of cloud-native and microservices, the API gateway has emerged as a critical component for connecting and managing various microservices. However, as the number of services continues to grow and changes occur more frequently, the traditional imperative configuration has become increasingly challenging to manage and maintain. GitOps, on the other hand, is an operational model that leverages version control systems and automated workflows. By supporting declarative configurations, GitOps provides a more efficient, reliable, and traceable mode of operation.

To enhance developing efficiency and operational reliability, APISIX has introduced a new tool that supports GitOps in a declarative manner. By embracing the declarative nature of GitOps, APISIX strengthens its integration with modern development and operational workflows. This integration enables smoother collaboration between developers and operations teams, promoting efficient and reliable management of the APISIX platform.

## Why Does APISIX Support GitOps Declarative Configuration

> Although APISIX offers a stand-alone mode that allows configuration through YAML files, it lacks seamless integration with related ecosystems such as CI/CD tools like Jenkins and ArgoCD. While the APISIX Ingress Controller project has made significant strides in this area, APISIX itself does not provide a comprehensive set of declarative tools to support GitOps when used in non-Kubernetes environments such as bare metal or virtual machines.

Reviewer: Tense?

> In traditional API gateway management, configurations and policies are typically manipulated using imperative methods, requiring manual modifications through command-line tools or management interfaces. This approach poses several challenges, including:

Reviewer: ```suggestion
In traditional API gateway management, configurations and policies are typically manipulated using imperative methods, requiring manual modifications through command-line tools or management interfaces. This approach poses several challenges:
```

- **Cumbersome Configuration Management**: Manual modifications of configurations are prone to errors, especially when dealing with large-scale gateways.

- **Poor Traceability**: Tracking the change history and version control of configurations becomes difficult.

- **Lack of Consistency**: Configuration discrepancies among multiple environments result in inconsistencies between development, testing, and production environments.

> The APISIX development team recognizes several benefits of supporting GitOps in a declarative manner to effectively address these challenges:

Reviewer: Also tense. Would it be better to use "recognized"?

1. **Improved Developer Efficiency**: By using GitOps with a declarative configuration approach, developers can directly manage API gateway configurations by modifying and committing configuration files in the code repository. This method aligns with the development workflow that developers are familiar with, reducing the learning curve and tool-switching costs, and thus enhancing developer productivity.

> 2. **Enhanced Operational Reliability**: Storing configuration files in a version control system ensures consistency and reliability. Each configuration change can be traced through its change history, and it becomes easy to roll back to a previous configuration state. This enables traceability, and auditability, and reduces the risk of failures caused by human errors.

Reviewer: ```suggestion
2. **Enhanced Operational Reliability**: Storing configuration files in a version control system ensures consistency and reliability. Each configuration change can be traced through its change history, and it becomes easy to roll back to a previous configuration state. This function enables traceability and auditability, and reduces the risk of failures caused by human errors.
```

## How Does APISIX Use ADC for Declarative Configuration

### Installing APISIX and ADC

Please refer to the [APISIX documentation](https://apisix.apache.org/docs/apisix/getting-started/README/) for installing APISIX. Once APISIX is installed, you can proceed to install the ADC binary to the `$GOPATH/bin` directory using the `go install` command.

```shell
go install github.com/api7/adc@latest
```

Add this line of code to your `$PATH` environment variable:

```shell
export PATH=$PATH:$GOPATH/bin
```

If you don't have Go installed, you can download the latest `adc` binary and add it to your `/bin` folder:

```shell
wget https://github.com/api7/adc/releases/download/v0.2.0/adc_0.2.0_linux_amd64.tar.gz
tar -zxvf adc_0.2.0_linux_amd64.tar.gz
mv adc /usr/local/bin/adc
```

You can find binaries for other operating systems on the [releases page](https://github.com/api7/adc/releases/tag/v0.2.0). In the future, these files will be published on package management tools like Homebrew.

Run the following code to confirm that `adc` has been installed:

```shell
adc --help
```

> If everything goes well, you will see a list of available subcommands and how to use them.

Reviewer: 这里“How to use them“是一个会pop up的section吗？如果不是，感觉还是改成”usage" or "using guide"之类的词会好一些。

### Configuring ADC with APISIX Instance

Next, configure the APISIX instance in the ADC.

```shell
adc configure
```

It will prompt you to pass in the APISIX server address ('http://127.0.0.1:9180' if you followed along) and token. If everything is filled in correctly, you can see the following content:

```shell
ADC configured successfully!
Connected to APISIX successfully!
```

We can use the `ping` subcommand to check the APISIX connection at any time:

```shell
adc ping
```

### Validating APISIX Configuration Files

Create a basic APISIX configuration with a route that forwards traffic to upstreams:

```yaml title="config.yaml"
name: "Basic configuration"
version: "1.0.0"
services:
  - name: httpbin-service
    hosts:
      - api7.ai
    upstream:
      name: httpbin
      nodes:
        - host: httpbin.org
          port: 80
          weight: 1
routes:
  - name: httpbin-route
    service_id: httpbin-service
    uri: "/anything"
    methods:
      - GET
```

> Once the ADC is connected to the running APISIX instance, we can use it to validate this configuration before applying it by running:

Reviewer: 感觉操作指南也是"we""you"混用着来，统一一下会不会更好呢？

```shell
creating service: "httpbin-service"
creating route: "httpbin-route"
Summary: created 2, updated 0, deleted 0
```

To verify that the route was created correctly, let's try sending a request:

```shell
curl localhost:9080/anything -H "host:api7.ai"
```

If everything is correct, we will receive a response from [httpbin.org](httpbin.org).

### Comparing Local and Running Configuration

Now, let's update the local configuration in the `config.yaml` file by adding another route:

```yaml title="config.yaml" {20-24}
name: "Basic configuration"
version: "1.0.0"
services:
  - name: httpbin-service
    hosts:
      - api7.ai
    upstream:
      name: httpbin
      nodes:
        - host: httpbin.org
          port: 80
          weight: 1
routes:
  - name: httpbin-route-anything
    service_id: httpbin-service
    uri: "/anything"
    methods:
      - GET
  - name: httpbin-route-ip
    service_id: httpbin-service
    uri: "/ip"
    methods:
      - GET
```

Before synchronizing this configuration with APISIX, ADC allows you to check the differences between it and the existing APISIX configuration. You can run the following operations:

```shell
adc diff -f config.yaml
```

> You can see the added and deleted configurations and understand the changes before applying the configuration.

Reviewer: ```suggestion
You can see the added and deleted configurations and check the changes before applying the configuration.
```

### Converting OpenAPI Definitions to APISIX Configurations

ADC also supports the use of [OpenAPI definitions](https://spec.openapis.org/oas/v3.0.0). ADC allows the conversion of OpenAPI format definitions into APISIX configurations.

> For example, if you documented your API in OpenAPI format like this:

Reviewer: ```suggestion
For example, if you document your API in OpenAPI format like this:
```

### Tip: Use Autocomplete

ADC offers many functions, and the list of features is sure to grow. To learn how to use any subcommand, you can use the `--help` or `-h` flag, which will display the documentation for that subcommand.

To make it even easier, you can use the `completion` subcommand to generate an autocompletion script for your shell environment. For example, if you are using a zsh shell, you can run:

```shell
adc completion zsh
```

You can then copy and paste the output into your `.zshrc` file and it will start showing hints when you use `adc`.

ADC is still in its infancy and is constantly being improved. To learn more about the project, report a bug, or suggest a feature, visit [github.com/api7/adc](github.com/api7/adc).

## Summary

By using the declarative configuration tool ADC, APISIX provides a more simplified, reliable, and traceable management method, allowing developers to manage and deploy API gateway configurations more efficiently. This new feature brings many benefits to team collaboration, environmental consistency, and configuration management, providing strong support for building reliable cloud-native architectures.

> Users can connect well with tools such as Jenkins and ArgoCD in non-Kubernetes environments, and use GitOps's internal CI/CD method to control various behaviors of APISIX to achieve functions such as multi-environment release.

Reviewer: ```suggestion
In non-Kubernetes environments, users can seamlessly integrate tools like Jenkins and ArgoCD. They can leverage GitOps' internal CI/CD approach to manage various aspects of APISIX, enabling functions like multi-environment releases.
```
